File: strings/check-if-a-string-contains-all-binary-codes-of-size-k/check-if-a-string-contains-all-binary-codes-of-size-k.go
package checkifastringcontainsallbinarycodesofsizek

func hasAllCodes(s string, k int) bool {
	if len(s) < k || len(s)-k+1 < (1<<k) {
		return false
	}

	seen := make(map[string]bool)

	total := 1 << k

	n := len(s)

	start := 0
	end := start + k

	for end <= n {
		curr := s[start:end]

		if _, ok := seen[curr]; !ok {
			seen[curr] = true
		}

		if len(seen) == total {
			return true
		}

		start++
		end++
	}
	return false
}

File: strings/check-if-a-string-contains-all-binary-codes-of-size-k/check-if-a-string-contains-all-binary-codes-of-size-k.js
/**
 * @param {string} s
 * @param {number} k
 * @return {boolean}
 */
function hasAllCodes(s, k) {
  if (s.length < k || s.length - k + 1 < 2 ** k) {
    return false;
  }

  const seen = new Set();

  const total = 2 ** k;

  const n = s.length;
  let start = 0;
  let end = start + k;

  while (end <= n) {
    const curr = s.substring(start, end);

    if (!seen.has(curr)) {
      seen.add(curr);
    }

    if (seen.size == total) {
      return true;
    }

    start++;
    end++;
  }
  return false;
}
